artifacts/oopsla25-bv-decide/Dockerfile

```diff
@@ -1,55 +1,79 @@
-	pip \
-	python3-venv \
+        pip \
+        python3-venv \
         build-essential \
         cmake \
         git \
         ninja-build \
-        gdb curl wget \
+        gdb curl wget zstd unzip sudo \
         # dependencies for experiment scripts
         python3-matplotlib python3-pandas python3-num2words \
         # bitwuzla dependencies
-        meson libgmp-dev libcadical-dev libsymfpu-dev pkg-config
+        meson libgmp-dev libcadical-dev libsymfpu-dev pkg-config \
+        # CoqQFBV dependencies
+        libboost-timer-dev mlton && \
+    apt-get clean
 
-# install Lean
-RUN curl https://raw.githubusercontent.com/leanprover/elan/master/elan-init.sh -sSf > /code/elan.sh \
-    && sh /code/elan.sh -y # elan
-ENV PATH /root/.elan/bin:$PATH
-ENV LIBRARY_PATH /root/.elan/lib:$LIBRARY_PATH
+# Create a new user named 'user' with no password and switch to it
+# Needed to install opam packages while avoiding sandboxing issues
+RUN useradd -m -s /bin/bash user && \
+    adduser user sudo && \
+    echo "user ALL=(ALL) NOPASSWD:ALL" >> /etc/sudoers
+RUN mkdir -p /code && chown user:user /code # Create /code as root before switching
 
-# install Bitwuzla (commit: 22.6.25)
-RUN git clone https://github.com/bitwuzla/bitwuzla && cd bitwuzla && git checkout 0dca38d0f62fa9002ad6278ca6374838a69ade19
-WORKDIR /code/bitwuzla
-RUN ./configure.py && cd build && ninja install
+USER user
 
-# install Lean-MLIR (commit: 22.06.25)
-RUN git clone https://github.com/opencompl/lean-mlir && cd lean-mlir && git checkout b0253c5f5dbc6cc507b559d76fd659da38a8e565 
-WORKDIR /code/lean-mlir
+# Install elan and update environment
+RUN curl https://elan.lean-lang.org/elan-init.sh -sSf | sh -s -- -y --default-toolchain none
+ENV PATH=/home/user/.elan/bin:$PATH
 
-# install plotting script dependencies
 
-# WORKDIR /code/lean-mlir/bv-evaluation/
-# RUN bash -c "\
-# 	python3 -m venv venv \
-# 	&& source venv/bin/activate \
-# 	&& python3 -m pip install -r /code/lean-mlir/bv-evaluation/requirements.txt "
+# Install Bitwuzla (commit: 22.6.25)
+RUN cd /home/user && git clone https://github.com/bitwuzla/bitwuzla
+# && git checkout 0dca38d0f62fa9002ad6278ca6374838a69ade19
+WORKDIR /home/user/bitwuzla
+RUN ./configure.py && cd build && ninja && sudo ninja install
 
-# get cached dependencies
+
+# Install Lean-MLIR (commit: 26.06.25)
+RUN cd /code && git clone https://github.com/opencompl/lean-mlir && cd lean-mlir
+#&& git checkout 7bd8c716a4a0bd5608279f0b73bd6c87f78c9f1c
 WORKDIR /code/lean-mlir
-COPY . .
 
-# build lean-mlir
-RUN lake exe cache get
-RUN lake build
+# Install and set the default Lean toolchain for elan
+# Get the toolchain from .lean-toolchain file (example: 'leanprover/lean4:nightly-2025-06-20')
+RUN /home/user/.elan/bin/elan toolchain install nightly && \
+    /home/user/.elan/bin/elan default nightly
+
+# Ensure the .lake/packages directory is clean before running lake update
+RUN rm -rf .lake/packages/mathlib && lake update && lake build
```

Comment:
> I am very unsure about the whole versioning of lean/lean-mlir/mathlib: I will need some help with that.

`lake` should automatically figure out the correct Lean toolchain version to use in a clean build with `lake update && lake build`. I don't know what happens when you start with an old build (which I am guessing you do from `rm -rf .lake/packages/mathlib`).

Reply:
I see, the problem is that without those lines kept getting an error in the docker, which was unable to find the `lean-toolchain` or `lakefile`

Reply:
Oh! That's interesting... It just works for me...

Reply:
let me give it another try